### PR TITLE
Fix some arithmetic invalidations

### PR DESCRIPTION
**Pull request by omus**
While investigating https://github.com/invenia/Intervals.jl/issues/144 I noticed these invalidations:

```julia
 inserting +(a, b::T) where T<:Interval in Intervals at /Users/omus/.julia/dev/Intervals/src/interval.jl:312 invalidated:
   mt_backedges: 1: signature Tuple{typeof(+), Ptr{UInt8}, Any} triggered MethodInstance for memcpy(::Ptr{UInt8}, ::Int64, ::Ptr{UInt8}, ::Any, ::Int64) (0 children)
                 2: signature Tuple{typeof(+), Ptr{UInt8}, Any} triggered MethodInstance for pointer(::String, ::Integer) (6 children)
```

The use of `Any` stuck out to me so I looked into these and the changes PR was is the result. With these changes this invalidations no longer exist. Please let me know if I'm off base here as I am new to addressing invalidations.

### CODE REVIEW

**Comment by omus** [comment]
Doctest failure looks related. Will investigate. Stacktrace:
```julia
ERROR: LoadError: TypeError: in typeassert, expected UInt64, got a value of type Int64
Stacktrace:
  [1] writeshortest(buf::Vector{UInt8}, pos::Int64, x::Float64, plus::Bool, space::Bool, hash::Bool, precision::Int64, expchar::UInt8, padexp::Bool, decchar::UInt8, typed::Bool, compact::Bool)
    @ Base.Ryu ./ryu/shortest.jl:387
  [2] show(io::IOContext{IOBuffer}, x::Float64, forceuntyped::Bool, fromprint::Bool)
    @ Base.Ryu ./ryu/Ryu.jl:115
  [3] show(io::IOContext{IOBuffer}, x::Float64)
...
```

File: base/ryu/shortest.jl
Comment (KristofferC) [inline base/ryu/shortest.jl:368]: Only adding a type assert usually doesn't help. You tend to either have to also add a conversion or figure out why this didn't concretely infer in the first place.

File: base/strings/string.jl
Comment (KristofferC) [inline base/strings/string.jl:95]: Sometimes these are written as `Int(i)::Int - 1`.

File: base/ryu/shortest.jl
Comment (omus) [reply base/ryu/shortest.jl:368]: Thanks for the guidance. It appears the invalidation with `memcpy` has disappeared so I'll revert these changes. I did find some other changes I may still add in

File: base/strings/string.jl
Comment (omus) [reply base/strings/string.jl:95]: Is there any concern about having values too large to convert to an `Int`. It seems unlikely that would occur but I figured I'd ask

Comment (timholy) [reply base/strings/string.jl]: I'd think string length is pretty safe; someone would need to have a single string that's more than half the size of memory on a 32-bit machine. I guess you could alternatively use `UInt`.

File: base/ryu/shortest.jl
Comment (KristofferC) [reply base/ryu/shortest.jl]: > It appears the invalidation with memcpy has disappeared 

Could have been https://github.com/JuliaLang/julia/pull/37802. I recall looking at this exact code for some invalidation and it is likely that I made #37802 in response to it, but I don't exactly recall.

**Comment by omus** [comment]
I think there is nothing left to do here. I'll wait to merge until I get approval

**Review by timholy — APPROVED** [review]

**Pull request merged**
